arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Option.kt

```diff
@@ -1290,32 +1301,31 @@ public fun <A> Option<Option<A>>.flatten(): Option<A> =
 @Deprecated(
   NicheAPI + "Prefer using the Option DSL or explicit map with orElse",
   ReplaceWith(
-    "map(fb).orElse { Some(fe(Unit)) }",
-    "arrow.core.Some",
-    "arrow.core.orElse"
+    "map(fb).recover { fe(Unit) }",
+    "arrow.core.recover"
   )
 )
 public inline fun <A, B> Option<A>.redeem(fe: (Unit) -> B, fb: (A) -> B): Option<B> {
   contract {
     callsInPlace(fe, InvocationKind.AT_MOST_ONCE)
     callsInPlace(fb, InvocationKind.AT_MOST_ONCE)
   }
-  return map(fb).handleError(fe)
+  return map(fb).recover { fe(Unit) }
 }
 
 @Deprecated(
   NicheAPI + "Prefer using the Option DSL or explicit flatMap with orElse",
   ReplaceWith(
-    "flatMap(fb).orElse(fe)",
-    "arrow.core.orElse"
+    "flatMap(fb).recover { fe(Unit).bind() }",
```

Comment:
This can also be replace with `fold({ fe(Unit) }, fb)`.

Reply:
I have been thinking about the original implementation for `redeemWith`, which doesn't match the `fold` behavior. Keeping in mind the current implementation:
```
flatMap(fb).handleErrorWith(fe)
```
If the function `fb` returns `None`, we can still recover and return the result of the function `fe`. However, if we use `fold`, we would return `None`. For example:
```
fun fe() = Some("recovered from None")
fun fb(a: A) = None

Some("hello").flatMap(fb).recover { fe(Unit).bind() } // Some("recovered from None")
Some("hello").fold({ fe(Unit) }, fb) // None
```
`redeem` does behave as the `fold` function, so maybe the original implementation was wrong, and we should fix it. What do you think, @nomisRev?

Reply:
The original inspiration comes from `MonadError` in Cats, which defines it like this: `fa.redeemWith(fe, fs) <-> fa.attempt.flatMap(_.fold(fe, fs))`. In this case `fe` cannot recover from a failure in `fe`. 

Which is in line with the original implementation (for `Either`):

```kotlin
  when (this) {
    is Left -> fa(this.value)
    is Right -> fb(this.value)
  }
```

However, I think we should not change the behavior of functions that are people are currently using in production. So we should keep the behavior the same for `Option` which was `flatMap(fb).handleErrorWith(fe)`.

This shows more proof that these APIs are a bit confusing, and reducing API surface will simplify Arrow a lot. `Option` and `Either` were not consistent in `redeemWith`.

So I am marking this comment as resolved, and the deprecation for `Option` should stay `flatMap(fb).recover { fe(Unit).bind() }` to remain consistent with current behavior.

Reply:
Yes, it makes sense to keep the original implementation, even if it doesn't follow the expected behavior. As you've mentioned, people might be using it in production as it is.